tests-integration/tests/crowdfund_app.rs

```diff
@@ -0,0 +1,573 @@
+
+    // Claim tier
+    let _ = crowdfund
+        .execute_claim(buyer_one.clone(), &mut router)
+        .unwrap();
+    // buyer_one should own 30 tiers now (10 pre order + 20 purchased)
+    let owner_resp = cw721.query_owner_of(&router, "0".to_string());
+    assert_eq!(owner_resp, buyer_one.to_string());
+    let owner_resp = cw721.query_owner_of(&router, "29".to_string());
+    assert_eq!(owner_resp, buyer_one.to_string());
+}
+
+#[rstest]
+fn test_failed_crowdfund_app_cw20(#[with(false)] setup: TestCase) {
+    let TestCase {
+        mut router,
+        andr,
+        crowdfund,
+        presale,
+        cw20,
+        ..
+    } = setup;
+    let cw20 = cw20.unwrap();
+
+    let owner = andr.get_wallet("owner");
+    let buyer_one = andr.get_wallet("buyer_one");
+
+    // Start campaign
+    let start_time = None;
+    let end_time = Milliseconds::from_nanos(router.block_info().time.plus_days(1).nanos());
+
+    let _ = crowdfund.execute_start_campaign(
+        owner.clone(),
+        &mut router,
+        start_time,
+        end_time,
+        Some(presale),
+    );
+    let summary = crowdfund.query_campaign_summary(&mut router);
+    assert_eq!(summary.current_cap, 0);
+    assert_eq!(summary.current_stage, CampaignStage::ONGOING.to_string());
+
+    // Purchase tiers
+    router.set_block(BlockInfo {
+        height: router.block_info().height,
+        time: router.block_info().time.plus_seconds(1),
+        chain_id: router.block_info().chain_id,
+    });
+
+    let orders = vec![SimpleTierOrder {
+        level: Uint64::one(),
+        amount: Uint128::new(5),
+    }];
+
+    let buyer_one_original_balance = cw20.query_balance(&router, buyer_one.clone());
+    let hook_msg = mock_purchase_cw20_msg(orders);
+
+    cw20.execute_send(
+        &mut router,
+        buyer_one.clone(),
+        crowdfund.addr(),
+        Uint128::new(5000),
+        &hook_msg,
+    )
+    .unwrap();
+
+    let buyer_one_balance = cw20.query_balance(&router, buyer_one.clone());
+
+    assert_eq!(
+        buyer_one_balance,
+        buyer_one_original_balance - Uint128::new(5 * 100)
+    );
+
+    // End campaign
+    router.set_block(BlockInfo {
+        height: router.block_info().height,
+        time: router.block_info().time.plus_days(2),
+        chain_id: router.block_info().chain_id,
+    });
+
+    let _ = crowdfund.execute_end_campaign(owner.clone(), &mut router);
+    let summary = crowdfund.query_campaign_summary(&mut router);
+    assert_eq!(summary.current_cap, 5 * 100);
+    assert_eq!(summary.current_stage, CampaignStage::FAILED.to_string());
+
+    // Refund
+    let buyer_one_original_balance = cw20.query_balance(&router, buyer_one.clone());
+    let _ = crowdfund
+        .execute_claim(buyer_one.clone(), &mut router)
+        .unwrap();
+    let buyer_one_balance = cw20.query_balance(&router, buyer_one.clone());
+    assert_eq!(
+        buyer_one_balance,
+        buyer_one_original_balance + Uint128::new(5 * 100)
+    );
+}
+
```

Comment:
Can we have one example of an ADO being the recipient of the crowdfund? Having a case where it sends to an ADO but fails (for whatever reason) and can be cancelled may also be nice but not mandatory.

Reply:
Test case added for that

Reply:
Mind adding a successful one too? Maybe to a splitter?

Reply:
🤔Not sure about the purpose of the successful one. Added a test case with valid recipient as well and checked if the funds are distributed with the splitter anway. Any specific tests might be needed for the successful one case?